staging/src/k8s.io/externaljwt/pkg/plugin/plugin.go

```diff
@@ -0,0 +1,263 @@
+}
+
+// GetServiceMetadata returns metadata associated with externalJWTSigner
+// It Includes details like max token lifetime supported by externalJWTSigner, etc.
+func (p *Plugin) GetServiceMetadata(ctx context.Context) (*Metadata, error) {
+	ctx, cancel := context.WithTimeout(context.Background(), 60*time.Second)
+	defer cancel()
+
+	req := &ejspb.MetadataRequest{}
+	resp, err := p.client.Metadata(ctx, req)
+	if err != nil {
+		return nil, fmt.Errorf("while getting metadata from external jwt signer: %w", err)
+	}
+
+	return &Metadata{
+		MaxTokenExpirationSeconds: int(resp.MaxTokenExpirationSeconds),
+	}, nil
+}
+
+// IsExternal informs if TokenGenerator intigerates with external-jwt-signer.
+func (p *Plugin) IsExternal() bool {
+	return true
+}
+
+// setKeyCache is for use with testcases.
+// It is required because creating keyCache itself requires a reference to Plugin.
+func (p *Plugin) setKeyCache(keyCache *KeyCache) {
+	p.keyCache = keyCache
+}
+
+func mergeClaims(claims *jwt.Claims, privateClaims interface{}) ([]byte, error) {
+	privateBytes, err := json.Marshal(privateClaims)
+	if err != nil {
+		return nil, fmt.Errorf("while marshalling private claims: %w", err)
+	}
+
+	var unifiedMap map[string]any
+	if err := json.Unmarshal(privateBytes, &unifiedMap); err != nil {
+		return nil, fmt.Errorf("while unmarshalling private claims: %w", err)
+	}
+
+	unifiedMap["iss"] = claims.Issuer
+	unifiedMap["sub"] = claims.Subject
+	unifiedMap["aud"] = claims.Audience
+	unifiedMap["exp"] = claims.Expiry
+	unifiedMap["nbf"] = claims.NotBefore
+	unifiedMap["iat"] = claims.IssuedAt
+	if claims.ID != "" {
+		unifiedMap["jti"] = claims.ID
+	}
+
+	unifiedBytes, err := json.Marshal(unifiedMap)
+	if err != nil {
+		return nil, fmt.Errorf("while marshalling unified claims: %w", err)
+	}
+
+	return unifiedBytes, nil
+}
+
+func (p *Plugin) validateJWTHeader(response *ejspb.SignJWTResponse) error {
+	jsonBytes, err := base64.RawURLEncoding.DecodeString(string(response.GetHeader()))
+	if err != nil {
+		return fmt.Errorf("while unwrapping header: %w", err)
+	}
+
+	decoder := json.NewDecoder(bytes.NewBuffer(jsonBytes))
+	decoder.DisallowUnknownFields()
+
+	header := &struct {
+		Algorithm string `json:"alg,omitempty"`
+		KeyID     string `json:"kid,omitempty"`
+		Type      string `json:"typ,omitempty"`
+	}{}
+
+	if err := decoder.Decode(header); err != nil {
+		return fmt.Errorf("while parsing header JSON: %w", err)
+	}
+
+	if header.Type != "JWT" {
+		return fmt.Errorf("bad type")
+	}
+	if header.KeyID == "" {
+		return fmt.Errorf("key id missing")
+	}
+	switch header.Algorithm {
+	case "RS256", "ES256", "ES384", "ES512":
+		// OK
+	default:
+		return fmt.Errorf("bad signing algorithm %q", header.Algorithm)
+	}
+
+	publicKeys := p.keyCache.GetPublicKeys(header.KeyID)
+
+	if len(publicKeys) == 0 {
+		return fmt.Errorf("key used for signing JWT (kid: %s) is not among supported keys", header.KeyID)
+	}
+
+	for _, key := range publicKeys {
```

Comment:
only do the oidc discovery validation if the token being issued is an oidc-compatible token (has audience, etc) (need to pass in whether to do that validation as a parameter)

if this plugin is used by a controller to generate legacy tokens, being signed by a key that sets ExcludeFromOidcDiscovery is fine

Reply:
I find this to be a strange request since we never wire that code that way.  Either way I would not do it as a param to the methods, but instead as a config on the `Plugin` itself.

Reply:
sure, that's fine as well

Reply:
Done